### ISSUE TITLE
batchInsert errors not propagated in a promise chain

### ISSUE THREAD

**Issue by joelmukuthu** [knex/knex#1668]
`knex.batchInsert` does not propagate errors properly, that is, it doesn't return a promise.

Consider this example: 

``` js
var knex = require('knex')({
   client: 'pg',
   connection: 'connection-string-here'
});

knex.schema.createTable('users', function (table) {
   table.string('username').unique();
}).then(function () {
   return knex.table('users').insert({
      username: 'foo'
   });
}).then(function () {
   // generate a "duplicate key violation" error
   return knex.batchInsert('users', [{
      username: 'foo'
   }]); // yields uncaught error
}).catch(function (e) {
   console.log('caught error', e);
});
```

As opposed to this one where the error is caught:

``` js
knex.schema.createTable('users', function (table) {
   table.string('username').unique();
}).then(function () {
   return knex.table('users').insert({
      username: 'foo'
   });
}).then(function () {
   // generate a "duplicate key violation" error
   return knex('users').insert([{
      username: 'foo'
   }]);
}).catch(function (e) {
   console.log('caught error', e);
});
```

Is `knex.batchInsert` meant to return a promise? Because in that regard, `knex.batchInsert().catch()` throws a `TypeError('knex.batchInsert(...).catch is not a function')` error. From [the code](https://github.com/tgriesser/knex/blob/2e1a459a9e740f24b9a4647bd4da427854e551dd/src/util/batchInsert.js), it seems to me that unless you call `knex.batchInsert().then()` you won't get a promise, which doesn't seem quite right.

**Comment by tgriesser** [comment]
Yeah, I hadn't looked closely at the batchInsert code when it was added but it could definitely use a little work. I'll try and clean up these issues soon

**Comment by mercmobily** [comment]
Thanks for reporting this issue about `batchInsert` error handling and promise behavior.

**Good news**: This issue has been resolved! The problem you described where `batchInsert` didn't properly return a promise and could lead to unhandled rejection errors was fixed in commit 50557f75 (#1984).

**What was fixed**:
1. `batchInsert` now always returns a proper Promise that can be chained with `.then()` and `.catch()`
2. Errors are properly propagated through the promise chain
3. The implementation was completely rewritten to use modern async/await patterns

**Current Implementation**: 
The modern `batchInsert` (located at lib/execution/batch-insert.js) now:
- Returns a Promise extended with additional methods (`returning()` and `transacting()`)
- Properly handles errors within transactions
- Ensures all errors are catchable in the promise chain
- Uses `Object.assign()` to create a thenable object that always behaves as a promise

Your example code would now work correctly - the duplicate key violation error would be caught by the `.catch()` handler as expected.

**Additional improvements since this issue**:
- Support for transactions via `.transacting()` method
- Support for returning columns via `.returning()` method  
- Better input validation for batch arrays and chunk sizes
- Improved error messages

To use the fixed version, make sure you're on Knex 0.12.9 or later (the fix was included in March 2017).

**Related issues:**
- #1880 - Original unhandled rejection issue
- #1984 - PR that fixed the promise handling
- #1992 - Additional bugfixes for MySQL/MariaDB transactions
- #1418 - Added transaction support to batchInsert
